### ISSUE TITLE
Possibility to add live streams of Ring Cameras 

### ISSUE THREAD

**Issue by timota** [tsightler/ring-mqtt#121]
Hello.

First of all - your work is awesome. 

Is it possible to add live streams for Ring Cameras ? at least any plans for this ? 

Recently found this addon (https://github.com/jeroenterheerdt/ring-hassio) that based on the same API as your script. But that addon very limited and luck of many features. 

Thanks in advance.

**Comment by tsightler** [comment]
I don't think this will be possible with this project at the current time.  The goal of this project is to provide an interface between Ring devices and MQTT, so that anything that speaks MQTT (pretty much any Home Automation platform) can interface with Ring devices.  Yes, I implemented Home Assistant style MQTT discovery, which makes it very easy to integrate into Home Assistant, thus why there is also an addon, but the core project itself (ring-mqtt) is not Home Assistant specific.

Unfortunately, MQTT is not really a platform to support live streaming and, as far as I know, there is no generic way to stream video via MQTT.  There is a Home Assistant MQTT component for cameras and it does have the ability to upload still images from cameras, and I've considered adding this, but that would be at most a picture every 30 seconds or so because that's all that Ring allows for battery powered devices, and there are other limitations as well (some don't allow snapshots while recoding video for example).  I'm just not convinced of the usefulness of such a feature because of this, especially since the native Home Assistant Ring integration allows similar capabilities.

However, nothing keeps you from using both addons at the same time, and I really don't see the advantage in combining them as they do completely different things, but if you have some ideas how the two might be used together, then I'd be interested to hear it.

**Comment by tsightler** [comment]
As this is out of scope for this project closing this issue for now.

**Comment by tpmullan** [comment]
Could mqtt simply pass a streaming url that is consumed by home assistant or another service?

**Comment by tsightler** [comment]
Live streams are SIP/RTP sessions.  Something has to manage the session, start the stream and publish it to a URL for consumption, then stop the stream on shutdown, etc.  Of course all of that could be done (the ring-hassio addon does basically this), but it's unclear to me why sending a URL via MQTT by itself is useful.

Management of SIP/RTP remains well outside of the scope of this project and I really think queries about live streaming are better directed at projects that have that as a goal.

**Comment by tpmullan** [comment]
That is fair, I thought that the connection would be something that could be consumed by homeassistant.  If the information is passed is such a way that the stream can be created by a client it could be useful.  Basically it sounds to me, that there would need to be support for a process in homeassistant to start the SIP/RTP session.  That should never be done within MQTT, but generic info / updating when tokens expire would be possible to send through MQTT.

Basically there is a lot of work that would need to be done on both sides to make something like that useful and from what I see there are already workarounds available.  Would something like this be useful for a generic camera object in MQTT or do you think this is too Ring specific to be useful?

**Comment by tsightler** [comment]
I feel like the native Home Assistant integration for Ring cameras already deals with this reasonably well.  I believe it currently only does motion videos rather than a livestream, but I see no reason it could not be extended to do live streaming and personally I think that's where the work should go as I believe at this point it's the only feature really missing for that integration.  

If I was really wanting this feature in Home Assistant I believe that's where I tried to add it, probably start with a custom component by branching the existing native component and add it there.  I don't believe it would take too much really, but it may not be a function available in the Python API the component uses, so that might be the highest barrier.  For me, the barrier is that I'm a worse Python programmer than I am Javascript, and that's a pretty sad statement.  Sure, I can do basic scripts and things with Python for admin style tasks, but anything above that would require me to elevate my Python game pretty significantly, and I don't have the personal time available to make that investment right now.

I do think Home Assistant can take RTSP streams for the generic camera component, so maybe it's easier than I think, but I believe it's pretty limited on the format and I'm not sure how it would be possible to coordinate creating the stream and getting that to lovelace, etc.  Most project seem to require manual configuration of the generic camera platform.  Maybe there's some role for this project to play there, like providing a service that can be called to initiate the stream and start the live stream, but I've not researched this.  Certainly if you just need a topic to trigger starting a session and passing the URL via another topic, that would be easily done.

Heck, at one point I considered doing livestream via MQTT since everything is local I thought it might not be too bad.  You can pretty easily convert the stream to MJPEG images and this works, but at the speed of about 1 frame per second.  Not sure that's really very useful.

**Comment by tpmullan** [comment]
Home assistant can take an rtsp stream.  The new updates to the nest component are doing that now. I think you are right and it would do really well in that python library, but I am not sure if anyone is actively updating that component. I personally use this project because it supports the security system, and the other ring integrations do not.

**Comment by tsightler** [comment]
If I ever just had tons of time, I'd probably play with it just for the learning experience.  Unfortunately, I don't see that happening any time in the coming months, but maybe one day.
